hubtools/src/lib.rs

```diff
@@ -652,23 +716,67 @@ impl RawHubrisArchive {
 
         let board = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("board")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
         let mut name = manifest
             .as_table()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .get("name")
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .as_str()
-            .ok_or(Error::BadTomlType)?
+            .ok_or(Error::BadTomlTypeBoard)?
             .to_owned();
 
+        // The optional epoch argument and optional app.toml epoch
+        // parameter both default to zero and must agree with each other.
+        let epoch_arg: u32 = if let Some(arg) = epoch {
+            arg.parse::<u32>()
+                .map_err(|_| Error::InvalidEpoch(arg.to_string()))?
+        } else {
+            0u32
+        };
+
+        let epoch_toml: u32 = manifest
+            .as_table()
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .get("epoch")
+            .ok_or(Error::BadTomlTypeEpoch)?
+            .as_integer()
+            .ok_or(Error::InvalidEpoch(format!("{}:{}", file!(), line!())))?
+            .try_into()
+            .map_err(|_| {
+                Error::InvalidEpoch(format!("{}:{}", file!(), line!()))
+            })?;
+
+        // Default to zero or the only non-zero value,
+        // otherwise, they need to match.
+        let epoc = match (epoch_arg, epoch_toml) {
+            (0, 0) => Ok(0),
+            (x, 0) => Ok(x),
+            (0, y) => Ok(y),
+            (x, y) => {
+                if x == y {
+                    Ok(x)
+                } else {
+                    Err(Error::InvalidEpoch(format!(
+                        "epoch value from app.toml({}) != arg({})",
+                        epoch_toml, epoch_arg
+                    )))
+                }
+            }
+        }?;
+        // EPOC in the caboose is a ascii decimal number representing a u32.
+        // A  missing EPOC tag is equivalent to EPOC="0".
+        // Anything else represents some future scheme or an error that will
+        // be rejected by the Hubris update_server.
+        let caboose_epoc = format!("{epoc}").to_owned();
```

Comment:
Did you run `clippy` on this file?  I feel like it would suggest

```suggestion
        let caboose_epoc = epoc.to_string();
```

Reply:
Clippy was happy with it. I am often running `cargo watch -c -- cargo clippy --all` in a window while editing.
But, you're right, that is better.